CICD: Release PR Creation Workflow and Slack Messaging

algobarb opened the pull request:
# Summary
- Adds a Release PR Creation github workflow
- Add Slack Messaging after a release branch completes CircleCI tests
- Add a script/bump_version.py to handle bumping version in setup.py 
- Support Hot-Fixes

## Testing
- The Description is populated with the changelog as expected in this PR: https://github.com/algorand/py-algorand-sdk/pull/495
- Slack message can be found in our internal #devops-experimental-notifications slack channel.

tzaffi commented on scripts/bump_version.py:
```diff
@@ -0,0 +1,28 @@
+import argparse
+import re
+
+parser = argparse.ArgumentParser(
```
To conform a bit more with the conventions of this repo (and python best practices) I recommend that all the code from line 7 be wrapped inside of a function:
```python
def bump_version():
    parser = argparse.ArgumentParser(
        description="updates the version for a release",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )
    ... etc. ....
```
and then at the bottom call this function with:

```python
if __name__ == "__main__":
    bump_version()
```

See for example `generate_init.py` in the same directory.

algobarb replied:
Thanks for the tips! I should follow python best practices.

algobarb replied:
Please take a look at the updated script. I'm open to suggestions to make it better.

algobarb replied:
oh wait hold on. I should run lint

tzaffi reviewed (commented):
Thanks for this wonderful addition! It looks great, but I had one suggestion regarding the python script.

tzaffi reviewed (approved):
LGTM

onetechnical commented on .github/workflows/create-release-pr.yml:
```diff
@@ -0,0 +1,127 @@
+                      --header "Accept: application/vnd.github.v3+json" \
+                      "https://api.github.com/repos/${{ github.repository }}/releases")
+          tag=$(echo "$releases" | jq -r '.[0].tag_name')
+          echo "latest-tag=$tag" >> $GITHUB_OUTPUT
```
Will the above only get published releases, not pre-releases? Or maybe we want that for comparison?

algobarb replied:
Updated with the releases/latest api and tested it here: https://github.com/algorand/py-algorand-sdk/commit/f064fa2d9f47580f0f1b531924fcb1b8b878e9a1 (will delete this branch later)

onetechnical commented on .github/workflows/create-release-pr.yml:
```diff
@@ -0,0 +1,127 @@
+        env:
+          CHANGELOG_CONTENT: ${{ steps.build-changelog.outputs.changelog }}
+          PREVIOUS_VERSION: ${{ steps.get-release.outputs.latest-tag }}
+          GH_TOKEN: ${{ github.token }}
```
Is this scoped to only work for release branches?

algobarb replied:
Well, the source_ref to create the release_branch can come from anywhere including a hot_fix branch. 

The parameter: ```source_ref:
        description: 'Source branch, commit, or tag to create the release branch from'```

algobarb replied:
Actually I removed the source_ref but I added a check to the "Create Release Branch if it does not exist" step which should ensure that we are on working on the release branch.

onetechnical reviewed (approved):
I had some questions and one nit but nothing blocking.

excalq commented on .github/workflows/create-release-pr.yml:
```diff
@@ -0,0 +1,127 @@
+          PREVIOUS_VERSION: ${{ steps.get-release.outputs.latest-tag }}
+        run: |
+          git branch --show-current
+          echo "$(tail -n +2 CHANGELOG.md)" > CHANGELOG.md
```
Verifying this should be `>` rather than `>>`?

excalq commented on .github/workflows/create-release-pr.yml:
```diff
@@ -0,0 +1,127 @@
+        type: string
+      release_version:
+        description: 'In the vX.X.X format and will be used for tag and branch name'
+        default: 'vx.x.x'
```
non-blocking: Is there any means of validating the presence of the `v` prefix? I'm anticipating `python3 scripts/bump_version.py ${RELEASE_VERSION:1}` doing something bad if it gets forgotten.

algobarb replied:
Yes, I set this input as a variable in line 17: 
```
env:
  RELEASE_VERSION: ${{ inputs.release_version }}
```
and it gets checked when it's first used in the second step:
```

      - name: Create Release Branch
        id: create-release-branch
        run: |
          if [[ $RELEASE_VERSION =~ ^v[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
```
If it doesn't match the `^v` then it will fail the build after checking out the code.

excalq replied:
Awesome, sorry I missed that!

excalq reviewed (approved):
Only the non-blocking question re version v-prefix, but great work here.

algobarb commented on .github/workflows/create-release-pr.yml:
```diff
@@ -0,0 +1,127 @@
+        with:
+          fromTag: ${{ env.PREVIOUS_VERSION }}
+          toTag: ${{ steps.create-release-branch.outputs.release-branch }}
+          ignorePreReleases: true
```
I fixed the other part where we got PREVIOUS_VERSION... so I think it's ok. But I guess it doesn't handle the case of when we have two pre-releases in a row...

algobarb commented:
Closing to test adding pre-release changes. Will re-open PR afterwards.

excalq commented on .github/workflows/create-release-pr.yml:
```diff
@@ -0,0 +1,189 @@
+        description: "Pre-Release version, e.g. 'b1', will be added behind the release_version in setup.py."
+        required: false
+        type: string
+      release_branch_exists:
```
If the step on line 66 created the release branch (if it doesn't exist) _prior_ to the checkout steps, would this still be necessary as an input?

excalq commented on .github/workflows/create-release-pr.yml:
```diff
@@ -0,0 +1,189 @@
+        with:
+          ref: ${{ env.RELEASE_BRANCH }}
+
+      - name: Create Release Branch
```
```suggestion
      - name: Create Release Branch if it does not exist
```

excalq commented on .github/workflows/create-release-pr.yml:
```diff
@@ -0,0 +1,189 @@
+          ref: ${{ env.RELEASE_BRANCH }}
+
+      - name: Create Release Branch
+        if: ${{ ! inputs.release_branch_exists }}
```
```suggestion
```

algobarb commented:
Thanks @excalq for the pointers!

I added the recommended changes as well as a few more changes in the rest of the yml file to remove the input parameters: **release_branch_exists** and **source_ref**:
- I added `fetch-option: 0` to the checkout step so it will fetch all of the branches rather than just one commit, so we can check if the release branch exists on remote. 
- I also added a condition that if the release branch exists, it will error if we aren't running the workflow from the release branch. It could indicate that someone may not have realized the release branch was already created. 
- I removed source_ref parameter and instead will use the "workflow_ref", i.e. the ref used to run the workflow.
